languages/elixir/exercises/concept/processes/.docs/after.md

```diff
@@ -0,0 +1,123 @@
+
+  ```elixir
+  pid = spawn(fn -> 2 + 2 end)
+  Process.alive?(pid)
+  # => false
+  ```
+
+## Messages
+
+Processes do not directly share information with one another. Processes _send messages_ to share data. This concurrency pattern is called the [Actor model][wiki-actor-model].
+
+- Send messages to a process using [send/2][kernel-send].
+
+  ```elixir
+  send(pid, :hello)
+  ```
+
+  - The message ends up in the recipient's _mailbox_ in the order that they are sent.
+  - `send` does not check if the message was received nor if the recipient is still alive.
+
+- A message can be of any type.
+- You can receive a message sent to the current process using [`receive/1`][kernel-receive].
+
+  - You need to pattern match on messages.
+  - `receive` waits until one message matching any given pattern is in the processes mailbox.
+    - By default, it waits indefinitely, but can be given a timeout using an `after` block.
+  - Read messages are removed from the process' mailbox. Unread messages will stay there indefinitely.
+    - Always write a catch-all `_` clause in `receive/1` to avoid running of out memory due to piled up unread messages.
+
+  ```elixir
+  receive do
+    {:ping, sender_pid} -> send(sender_pid, :pong)
+    _ -> nil
+  after
+    5000 ->
+      {:error, "No message in 5 seconds"}
+  end
+  ```
+
+## Receive loop
+
+If you want to receive more than one message, you need to call `receive/1` recursively. It is a common pattern to implement a recursive function, for example named `loop`, that calls `receive/1`, does something with the message, and then calls itself to wait for more messages. If you need to carry some state from one `receive/1` call to another, you can do it by passing an argument to that `loop` function.
+
+```elixir
+def loop(state) do
+  receive do
+    :increment_by_one ->
+      loop(state + 1)
+
+    {:report_state, sender_pid} ->
+      send(sender_pid, state)
+      loop(state)
+
+    :stop ->
+      nil
+
+    _ ->
+      loop(state)
+  end
+end
+```
+
+In practice, this approach is rarely used directly. Elixir offers concurrency abstractions, such as the [`Agent` module][agent] or a [_`GenServer` behaviour`_][genserver], that both build on top of the receive loop. However, it is crucial to understand those basics to be able to efficiently use the abstractions.
+
+## PIDs
+
+- PIDs are their own datatype.
+  - You can check if a variable is a PID with [`is_pid/1`][kernel-is-pid]
+- Each process has its own unique PID.
+- You can get the current process' PID with `self()`.
+- PIDs function as _mailbox addresses_ - if you have a process' PID, you can send a message to that process.
+- PIDs are usually created indirectly, as a return value of functions that create new processes, like `spawn`.
+  - PIDs should not be created directly by the programmer. If it were required, Erlang has a [`list_to_pid/1`][erlang-list-to-pid] function.
+    ```elixir
+    :erlang.list_to_pid('<0.10.0>')
```

Comment:
@iHiD Is it allowed to have a snippet in `after.md` that depends on a concept that wasn't taught yet? Elixir doesn't have its own way (1) to create a PID directly, you need to do it using an Erlang's function, but calling Erlang functions from Elixir wasn't covered yet. If it's not allowed, I would just remove this point about PIDs.

(1) that's almost true, there is a helper for iex, but iex is another forbidden topic

Reply:
Yeah - I think this is totally fine :) Maybe say "You will learn more about this in the XYZ exercise" if you like.

Reply:
If you keep things minimal and just note that there _is_ a method/function to do X without explain X, I agree that it could work.